src/Database/LSMTree/Internal/Run.hs

```diff
@@ -156,18 +156,35 @@ readBlob fs Run {..} BlobSpan {..} = do
 -- TODO: Once snapshots are implemented, files should get removed, but for now
 -- we want to be able to re-open closed runs from disk.
 close :: HasFS IO h -> HasBlockIO IO h -> Run (FS.Handle h) -> IO ()
-close fs _hbio Run {..} = do
+close fs hbio Run {..} = do
     FS.hClose fs runKOpsFile
       `finally`
         FS.hClose fs runBlobFile
+    mapM_ (FS.hDropCacheAll' fs hbio) (pathsForRunFiles runRunFsPaths)
 
 -- | Should this run cache key\/ops data in memory?
 data RunDataCaching = CacheRunData | NoCacheRunData
   deriving stock Eq
 
+isNoCacheRunData :: RunDataCaching -> Bool
+isNoCacheRunData CacheRunData   = False
+isNoCacheRunData NoCacheRunData = True
+
+instance NFData RunDataCaching where
+  rnf CacheRunData   = ()
+  rnf NoCacheRunData = ()
+
 setRunDataCaching :: HasBlockIO IO h -> FS.Handle h -> RunDataCaching -> IO ()
-setRunDataCaching _ _ CacheRunData = return ()
-setRunDataCaching hbio runKOpsFile NoCacheRunData =
+setRunDataCaching hbio runKOpsFile CacheRunData = do
+    -- use normal readahead window (only applies to this file descriptor)
+    FS.hAdviseAll hbio runKOpsFile FS.AdviceNormal
+    -- use the page cache for disk I/O
+    FS.hSetNoCache hbio runKOpsFile False
+setRunDataCaching hbio runKOpsFile NoCacheRunData = do
+    -- drop the file from the OS page cache, disable file readahead (only
+    -- applies to this file descriptor)
+    FS.hAdvisesAll hbio runKOpsFile [FS.AdviceDontNeed, FS.AdviceRandom]
+    -- do not use the page cache for disk I/O
```

Comment:
Correct me if I'm wrong, but isn't this _only_ used for the kops file when used for reading pages? And so this is never called for reading or writing kops files during merge.

If so, then I think what we want is simpler:
* In the `CacheRunData` case, use just `AdviceRandom`
* In the `NoCacheRunData` case, use just `FS.hSetNoCache` with `True`.

We don't need to set `AdviceNormal` or `hSetNoCache` with `False`  since that's the default and a no-op.

We only need `AdviceRandom` when we _are_ using the page cache (i.e. not using `O_DIRECT`).

I don't think using `AdviceDontNeed` makes sense here. The best place to use that is after reading sequentially (e.g. after checking checksums during load).

Reply:
`setRundataCaching` is called both after opening/verification from disk, or after a merge has produced a new run. That is, it is called in `fromMutable`. I can try to move the `AdviceDontNeed` into the load/merge functions, but the effect would be the same

Reply:
With respect to the other suggestions: makes sense